cpp/src/arrow/compute/kernels/filter.cc

```diff
@@ -15,44 +15,453 @@
+        ++filtered_i;
+        continue;
+      }
+      if (!filter.Value(i)) {
+        continue;
+      }
+      if (values.IsNull(i)) {
+        null_bitmap_builder.UnsafeAppend(false);
+        offset_builder.UnsafeAppend(offset);
+        ++filtered_i;
+        continue;
+      }
+      null_bitmap_builder.UnsafeAppend(true);
+      value_slices[filtered_i] = list_array.value_slice(i);
+      ++filtered_i;
+      offset += list_array.value_length(i);
+      offset_builder.UnsafeAppend(offset);
+    }
+
+    std::shared_ptr<Array> out_values;
+    if (length != 0) {
+      RETURN_NOT_OK(Concatenate(value_slices, ctx->memory_pool(), &out_values));
+    } else {
+      out_values = list_array.values()->Slice(0, 0);
+    }
+
+    auto null_count = null_bitmap_builder.false_count();
+    std::shared_ptr<Buffer> offsets, null_bitmap;
+    RETURN_NOT_OK(offset_builder.Finish(&offsets));
+    RETURN_NOT_OK(null_bitmap_builder.Finish(&null_bitmap));
+
+    *out = MakeArray(ArrayData::Make(type_, length, {null_bitmap, offsets},
+                                     {out_values->data()}, null_count));
+    return Status::OK();
+  }
+};
+
+class DictionaryFilterImpl : public FilterKernel {
+ public:
+  DictionaryFilterImpl(const std::shared_ptr<DataType>& type,
+                       std::unique_ptr<FilterKernel> impl)
+      : FilterKernel(type), impl_(std::move(impl)) {}
+
+  Status Filter(FunctionContext* ctx, const Array& values, const BooleanArray& filter,
+                std::shared_ptr<Array>* out) override {
+    auto dict_array = checked_cast<const DictionaryArray*>(&values);
+    // To filter a dictionary, apply the current kernel to the dictionary's indices.
+    std::shared_ptr<Array> taken_indices;
+    RETURN_NOT_OK(impl_->Filter(ctx, *dict_array->indices(), filter, &taken_indices));
+    return DictionaryArray::FromArrays(values.type(), taken_indices,
+                                       dict_array->dictionary(), out);
+  }
+
+ private:
+  std::unique_ptr<FilterKernel> impl_;
+};
+
+class ExtensionFilterImpl : public FilterKernel {
+ public:
+  ExtensionFilterImpl(const std::shared_ptr<DataType>& type,
+                      std::unique_ptr<FilterKernel> impl)
+      : FilterKernel(type), impl_(std::move(impl)) {}
+
+  Status Filter(FunctionContext* ctx, const Array& values, const BooleanArray& filter,
+                std::shared_ptr<Array>* out) override {
+    auto ext_array = checked_cast<const ExtensionArray*>(&values);
+    // To take from an extension array, apply the current kernel to storage.
+    std::shared_ptr<Array> taken_storage;
+    RETURN_NOT_OK(impl_->Filter(ctx, *ext_array->storage(), filter, &taken_storage));
+    *out = ext_array->extension_type()->MakeArray(taken_storage->data());
+    return Status::OK();
+  }
+
+ private:
+  std::unique_ptr<FilterKernel> impl_;
+};
+
+template <typename T>
+using ArrayType = typename TypeTraits<T>::ArrayType;
+
+struct UnpackValues {
```

Comment:
I find this visitor-kernel-factory convoluted. It took me a while (and I already reviewed this in the past) to figure out that it "just" returns a kernel. The `FilterKernel::Make` would be much simpler if you'd just switch & cast. Sometimes repetition is better.

Reply:
I can refactor to a switch

Reply:
It could be fine with a visitor, I think the confusion comes from the usage of the Param class where it's both for parameters and return type.

Maybe just have a `Finalize` function like other visitors?